lib/ecto/adapters/sql.ex

```diff
@@ -458,43 +449,85 @@ defmodule Ecto.Adapters.SQL do
     end
   end
 
+  defp do_transaction(repo, opts, fun) do
+    {pool_mod, pool, timeout} = repo.__pool__
+    opts    = Keyword.put_new(opts, :timeout, timeout)
+    timeout = Keyword.fetch!(opts, :timeout)
+
+    case Pool.open_transaction(pool_mod, pool, timeout) do
+      {:ok, ref, {mod, _conn}, queue_time} ->
+        try do
+          mode = transaction_mode(pool_mod, pool, timeout)
+          transaction(repo, ref, mod, mode, queue_time, timeout, opts, fun)
+        catch
+          # If any error leaked, it was inside Ecto.
+          # Make sure we break the connection.
+          kind, reason ->
+            stack = System.stacktrace()
+            Pool.break(ref)
```

Comment:
`break/2` does a blocking disconnect with timeout and a checkin. `close_transaction/2` should come in an `else` clause rather than `after`. Is this misleading? The current behaviour is used so that checkin can occur as early as possible - the connection is useless once broken and we want to release the lock on the worker as soon as possible.

In #712 `break/2` is asynchronous because the worker checks itself in so the back pressure is not required. It is also more efficient (in that case) to send it as a single message.